core/drlibc/drlibc.c

```diff
@@ -164,6 +155,32 @@ clear_icache(void *beg, void *end)
     __asm__ __volatile__("isb" : : : "memory");
 #    endif
 }
+
+void
+set_cache_line_size_using_ctr_el0(OUT size_t *dcache_line_size,
+                                  OUT size_t *icache_line_size)
+{
+#    ifndef DR_HOST_NOT_TARGET
+    static size_t cache_info = 0;
+
+    /* "Cache Type Register" contains:
+     * CTR_EL0 [31]    : 1
+     * CTR_EL0 [19:16] : Log2 of number of 4-byte words in smallest dcache line
+     * CTR_EL0 [3:0]   : Log2 of number of 4-byte words in smallest icache line
+     * https://developer.arm.com/docs/ddi0595/h/aarch64-system-registers/ctr_el0
+     *
+     * Also, the whitepaper below documents AArch64 words being 32 bits wide.
+     * https://developer.arm.com/-/media/Files/pdf/
+     * graphics-and-multimedia/Porting%20to%20ARM%2064-bit.pdf
+     */
+    if (cache_info == 0)
+        __asm__ __volatile__("mrs %0, ctr_el0" : "=r"(cache_info));
+    if (dcache_line_size != NULL)
+        *dcache_line_size = 4 << (cache_info >> 16 & 0xf);
+    if (icache_line_size != NULL)
+        *icache_line_size = 4 << (cache_info & 0xf);
+#    endif
```

Comment:
It just does nothing for host!=target?  No return value?  That is not what a caller is going to expect: seems like it will lead to uninitialized variables and subsequent hard to debug bugs.

Reply:
I modified it to return false when no value is set, and asserted the return value to be true in existing call at `clear_icache`.

Reply:
Stepping back, is having this leaf function fail the right solution?  For x86-on-a64 (not yet fully supported), won't we have a similar issue where proc_init runs `cpuid`?  Shouldn't the solution be at a higher level?

This goes back to the earlier discussion.  Maybe all our current uses case should use the host.   For raw2trace: asking for the cache line size is a fallback and is expected to be the host.  Unfortunately it's a complex topic: things like proc_num_simd_registers(), if you're trying to generate some cross-arch instrumentation, you'd want to be for the target.

Reply:
Since this is a complex, unclear topic, for now let's go with your solution but with a XXX comment pointing at #1684.

Reply:
Done. Added the comment.